replay/models/nmslib_hnsw.py

```diff
@@ -0,0 +1,646 @@
+import logging
+import os
+import shutil
+import tempfile
+import weakref
+from typing import Any, Dict, Optional, Iterator, Union
+
+import nmslib
+import numpy as np
+import pandas as pd
+from pyarrow import fs
+from pyspark import SparkFiles
+from pyspark.sql import DataFrame
+from pyspark.sql import SparkSession
+from pyspark.sql.functions import pandas_udf
+from scipy.sparse import csr_matrix
+
+from replay.ann.ann_mixin import ANNMixin
+from replay.ann.utils import (
+    save_index_to_destination_fs,
+    load_index_from_source_fs,
+)
+from replay.session_handler import State
+from replay.utils import FileSystem, get_filesystem, FileInfo
+
+logger = logging.getLogger("replay")
+
+INDEX_FILENAME = "nmslib_hnsw_index"
+
+
+# pylint: disable=too-few-public-methods
+class NmslibIndexFileManager:
+    """Loads index from hdfs, local disk or SparkFiles dir and keep it in a memory.
+    Instance of `NmslibIndexFileManager` broadcasts to executors and is used in pandas_udf.
+    """
+
+    def __init__(
+        self,
+        index_params,
+        index_type: str,
+        index_file: Union[FileInfo, str]
+    ) -> None:
+
+        self._method = index_params["method"]
+        self._space = index_params["space"]
+        self._ef_s = index_params.get("efS")
+        self._index_type = index_type
+        self._index_file = index_file
+        self._index = None
+
+    @property
+    def index(self):
+        """Loads `nmslib hnsw` index from local disk, hdfs or spark files directory and returns it.
+        Loads the index only on the first call, then the loaded index is used.
+
+        :return: `hnswlib` index
+        """
+        if self._index:
+            return self._index
+
+        if self._index_type == "sparse":
```

Comment:
I would suggest to convert `if self._index_type == "sparse":` to one or more classes with inheritance, different classes for sparse/dense implementation

Reply:
removed the nmslib index construction for dense vectors, since there is the implementation for this via hnswlib